### PR TITLE
Refactor device registry, split-up connection and services

### PR DESCRIPTION
**Pull request by ctron**
### Type of change

- Enhancement / new feature
- Refactoring

### Description

Restructure the deployment of the device registry. This allows separate adapter and management instances, as well as different backends for the device registry and device connection services (e.g. JDBC + Infinispan).

The motivation for this is:
* To be able to independently scale up:
  ** The device connection service
  ** The device registry service for the protocol adapters
  ** The device registry service for the management console
* To be able to use different backends (Infinispan, or different JDBC databases) for the device connection and the device registry service
* To be able to scale down the device management to zero, and still serve requests for the protocol adapters (e.g maintenance, backup, lockdown)

This will also

* Drop the file base device registry (as it was flawed from the beginning)
* Make it necessary to have deploy an external service (e.g. database or infinispan) before deploying the IoT infrastructure
* Require to run one more service (device connection service)

### Checklist

- [ ] Update/write design documentation in `./documentation/design`
- [x] Write tests and make sure they pass
- [x] Update documentation
- [x] Check RBAC rights for Kubernetes / OpenShift roles
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

### CODE REVIEW

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.registry.PostgresTreeDeviceRegistryTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.registry.PostgresFlatDeviceRegistryTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.registry.PostgresFlatDeviceRegistryTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.registry.PostgresFlatDeviceRegistryTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.registry.PostgresFlatDeviceRegistryTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.registry.H2DeviceRegistryTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.registry.H2DeviceRegistryTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.registry.PostgresTreeDeviceRegistryTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.registry.H2DeviceRegistryTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.registry.H2DeviceRegistryTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.registry.H2DeviceRegistryTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.registry.H2DeviceRegistryTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.registry.H2DeviceRegistryTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.registry.H2DeviceRegistryTest#testDeviceCredentialsPlainPassword

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.registry.H2DeviceRegistryTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.registry.H2DeviceRegistryTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.registry.H2DeviceRegistryTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.registry.PostgresTreeDeviceRegistryTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.registry.InfinispanDeviceRegistryTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.registry.InfinispanDeviceRegistryTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.registry.InfinispanDeviceRegistryTest

**Comment by kornys** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.registry.InfinispanDeviceRegistryTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.IoTProjectManagedTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.IoTProjectManagedTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.IoTProjectManagedTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.registry.PostgresTreeDeviceRegistryTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.project.ManagedTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.project.ManagedTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.MultipleProjectsTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.MultipleProjectsTest

File: api-model/src/main/java/io/enmasse/iot/model/v1/DeviceConnectionServiceConfig.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2019, EnMasse authors.
```
Comment (lulf) [inline api-model/src/main/java/io/enmasse/iot/model/v1/DeviceConnectionServiceConfig.java]: 2020

File: api-model/src/main/java/io/enmasse/iot/model/v1/ExternalJdbcRegistryServer.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2019, EnMasse authors.
```
Comment (lulf) [inline api-model/src/main/java/io/enmasse/iot/model/v1/ExternalJdbcRegistryServer.java]: 2020

**Comment by ctron** [comment]
@lulf I updated the initial PR comment, trying to answer your questions.

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.isolated.MultipleProjectsTest

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot

File: pkg/controller/iotconfig/infinispan_device_connection.go [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+  iot:
+
+    app:
+      maxInstances: 1
+
+    vertx:
+      preferNative: true
+
+    healthCheck:
+      insecurePortBindAddress: 0.0.0.0
```
Comment (lulf) [inline pkg/controller/iotconfig/infinispan_device_connection.go:223]: Should this be 127.0.0.1?

File: pkg/controller/iotconfig/jdbc_device_connection.go [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+			return err
+		}
+
+		container.Args = nil
+
+		// set default resource limits
+
+		container.Resources = corev1.ResourceRequirements{
+			Limits: corev1.ResourceList{
+				corev1.ResourceMemory: *resource.NewQuantity(512*1024*1024 /* 512Mi */, resource.BinarySI),
```
Comment (lulf) [inline pkg/controller/iotconfig/jdbc_device_connection.go:80]: Should this be overridable somewhere?

File: pkg/controller/iotconfig/jdbc_device_connection.go [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+		"enmasse": map[string]interface{}{
+			"iot": map[string]interface{}{
+				"app": map[string]interface{}{
+					"maxInstances": 1,
+				},
+				"vertx": map[string]interface{}{
+					"preferNative": true,
+				},
+				"healthCheck": map[string]interface{}{
+					"insecurePortBindAddress": "0.0.0.0",
```
Comment (lulf) [inline pkg/controller/iotconfig/jdbc_device_connection.go:224]: Should this be 127.0.0.1?

File: pkg/controller/iotconfig/jdbc_device_connection.go (hunk 4 above)
Comment (ctron) [reply pkg/controller/iotconfig/jdbc_device_connection.go:80]: This is just the default, the override takes place a bit further down below: https://github.com/EnMasseProject/enmasse/pull/4142/files/55e756995990b50ebef0dcb2d0826e7b736969c6#diff-b1e319091064da1010902dc38f8a3c7dR134 

~~~go
// apply container options

if service.JDBC != nil {
  applyContainerConfig(container, service.JDBC.Container)
}
~~~

File: pkg/controller/iotconfig/infinispan_device_connection.go (hunk 3 above)
Comment (ctron) [reply pkg/controller/iotconfig/infinispan_device_connection.go:223]: This needs to be accessed by the health probe and the prometheus scraper. From what I remember that didn't work when it was bound to 127.0.0.1.

File: pkg/controller/iotconfig/jdbc_device_connection.go (hunk 5 above)
Comment (ctron) [reply pkg/controller/iotconfig/jdbc_device_connection.go:224]: This needs to be accessed by the health probe and the prometheus scraper. From what I remember that didn't work when it was bound to 127.0.0.1.

File: iot/iot-device-registry-jdbc/src/main/java/io/enmasse/iot/registry/jdbc/config/MetricsConfiguration.java [hunk 6]
```diff
@@ -23,11 +27,24 @@
      * @return The new meter registry customizer.
      */
     @Bean
-    public MeterRegistryCustomizer<MeterRegistry> commonTags() {
+    @Profile({PROFILE_REGISTRY_ADAPTER, PROFILE_REGISTRY_MANAGEMENT})
```
[the remark this answers is not in the document]
Comment (ctron) [reply iot/iot-device-registry-jdbc/src/main/java/io/enmasse/iot/registry/jdbc/config/MetricsConfiguration.java:30]: I was thinking about that. There is a bunch of Spring components, which could go to some "base" module. However, I wanted to wait for the Hono API changes, and do this in another PR.

**Comment by ctron** [comment]
> I think you need to add the new service account and permissions to the olm-manifest/src/main/resources/latest/enmasse.clusterserviceversion.yaml

Thanks. Yes. I updated the file.

> Also, I think this is including changes from #4204 ?

It does, I would rebase this PR on master one #4204 merged.

File: iot/iot-device-registry-jdbc/src/main/java/io/enmasse/iot/registry/jdbc/config/MetricsConfiguration.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (dejanb) [reply iot/iot-device-registry-jdbc/src/main/java/io/enmasse/iot/registry/jdbc/config/MetricsConfiguration.java:30]: Sounds good

**Comment by ctron** [comment]
@enmasse-ci run tests profile=iot

**Comment by ctron** [comment]
Ok, the tests are fine. The only test failure was properly identified by the report portal as an "automation issue". It is the issue which is targeted to be fixed by #4227.

**Pull request merged**